Update ATen internals to use int64_t for dimension indexing

mike9ant opened the pull request:
Summary:
Some code ATen locations seemed to use int, etc. incorrectly where either int64_t or size_t was required. Update them to use int64_t for dimension indexing where necessary.

Differential Revision: D13950124

gchanan commented on aten/src/ATen/ExpandUtils.cpp:
```diff
@@ -8,12 +8,12 @@ std::vector<int64_t> infer_size(IntList a, IntList b) {
   ptrdiff_t ndim = dimsA > dimsB ? dimsA : dimsB;
```
I don't think this needs to be ptrdiff_t...there are literally no pointers involved here.

gchanan commented on aten/src/ATen/core/type.cpp:
```diff
@@ -25,7 +25,7 @@ std::ostream& operator<<(std::ostream & out, const Type & t) {
   } else if (auto value = t.cast<TensorType>()) {
     out << toString(value->scalarType()) << "(";
-    for (int i = 0; i < value->dim(); ++i) {
+    for (auto i = 0; i < value->dim(); ++i) {
```
this gets inferred as `int` btw, you probably want int64_t here.

gchanan commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -36,12 +36,12 @@ static inline Tensor integer_upcast(const Tensor& self, optional<ScalarType> dty
     mask.flip();
   } else {
-    for (int dim : dims) {
+    for (size_t dim : dims) {
```
this doesn't look right, shouldn't it be `int64_t` because that's the element type of `IntList`?

mike7ant replied:
Good catch, the fact that the vector's length is size_t while it holds int64_t is tripping me up :). Will update this and the others

gchanan commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -22,14 +22,14 @@ void TensorIterator::reorder_dimensions() {
   // returns 1 if the dim0 should come after dim1, -1 if dim0 should come
   // before dim1, and 0 if the comparison is ambiguous.
-  auto should_swap = [&](int dim0, int dim1) {
+  auto should_swap = [&](size_t dim0, size_t dim1) {
```
shouldn't these be `int64_t` because that's whats passed in?

mike7ant replied:
The situation here is just bad, because indexing of stride_bytes[dim0], etc. can only use size_t and technically not int64_t (although perhaps it practice doesn't matter if this isn't ran on 32-bit systems).

All of TensorIterator looks to not be typed properly, per our discussion. Perhaps we need dedicated typedef for dimension size type... so will leave this for later.

gchanan commented:
also CC @colesbury who occasionally used int, arguing that it's faster and that we don't actually support number of dimensions that big...

mike9ant committed:
Update ATen internals to use int64_t for dimension indexing (#16739)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/16739

Some code ATen locations seemed to use int, etc. inclorrectly where either
int64_t or size_t was required. Update them to use int64_t for dimension indexing where necessary.

Reviewed By: ezyang

Differential Revision: D13950124

fbshipit-source-id: 6ccc4366d1f5757152709944cf2d82d7b327d0f0
